### PR TITLE
Alicloud MNS service Pub/Sub component

### PR DESCRIPTION
**Pull request by cshi0**
# Description

Added a pub/sub component that supports Alibaba cloud MNS serivice

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #961

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#1573

### CODE REVIEW

File: pubsub/alicloud/mns/settings.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	// use one-to-one mode for messaging
+	MnsModeQueue = "queue"
+	// use one-to-many mode for messaging
+	MnsModeTopic = "topic"
+)
+
+// MNS settings
+type Settings struct {
+	// url for mns service
+	Url string `mapstructure:"url"`
```
Comment (pkedy) [inline pubsub/alicloud/mns/settings.go:27]: Nit: Go naming `Url` -> `URL`

File: pubsub/alicloud/mns/settings.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	// mns access key id
+	AccessKeyId string `mapstructure:"accessKeyId"`
```
Comment (pkedy) [inline pubsub/alicloud/mns/settings.go:29]: Nit: Go naming `AccessKeyId` -> `AccessKeyID`

File: pubsub/alicloud/mns/settings.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	// mns access key secret
+	AccessKeySecret string `mapstructure:"accessKeySecret"`
+	// mns token, optional
+	Token string `mapstructure:"token"`
+	// timeout in seconds, default 35
+	TimeoutSecond int64 `mapstructure:"timeoutSecond" default:"35"`
+	// mns mode (queue or topic)
+	MnsMode string `mapstructure:"mnsMode"`
```
Comment (pkedy) [inline pubsub/alicloud/mns/settings.go:37]: Nit: Go naming `MnsMode` -> `MNSMode` (assuming MNS is an acronym)

File: pubsub/alicloud/mns/settings.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+	ali_mns "github.com/aliyun/aliyun-mns-go-sdk"
+	"github.com/creasty/defaults"
+	"github.com/dapr/components-contrib/internal/config"
+)
+
+var (
+	// for more info: https://www.alibabacloud.com/help/doc-detail/27414.htm?spm=a2c63.p38356.879954.3.61943078qdiiEF#section-ghy-14s-7xd
+	// use one-to-one mode for messaging
+	MnsModeQueue = "queue"
```
Comment (pkedy) [inline pubsub/alicloud/mns/settings.go:19]: Nit: Go naming `MnsModeQueue` -> `MNSModeQueue` (assuming MNS is an acronym)

File: pubsub/alicloud/mns/settings.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	// use one-to-many mode for messaging
+	MnsModeTopic = "topic"
```
Comment (pkedy) [inline pubsub/alicloud/mns/settings.go:21]: Nit: Go naming `MnsModeTopic` -> `MNSModeTopic` (assuming MNS is an acronym)

File: pubsub/alicloud/mns/mns_test.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+	"time"
+
+	ali_mns "github.com/aliyun/aliyun-mns-go-sdk"
+	"github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/kit/logger"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMNSQueue(t *testing.T) {
+	mns := NewMns(logger.NewLogger("mns-test"))
```
Comment (pkedy) [inline pubsub/alicloud/mns/mns_test.go:16]: Nit: Go naming `NewMns` -> `NewMNS` (assuming MNS is an acronym)

File: pubsub/alicloud/mns/mns.go [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+	var msgSendReq ali_mns.MessageSendRequest
+	var msgPublishReq ali_mns.MessagePublishRequest
+	var mailAttr ali_mns.MailAttributes
+
+	json.Unmarshal(jsonBody, &msgSendReq)
+	json.Unmarshal(jsonBody, &msgPublishReq)
+	json.Unmarshal(jsonBody, &mailAttr)
+
+	delaySeconds, _ := strconv.ParseInt(req.Metadata["delay_seconds"], 10, 64)
+	msgSendReq.DelaySeconds = delaySeconds
```
Comment (pkedy) [inline pubsub/alicloud/mns/mns.go:99]: Is `0` a valid value for `DelaySeconds`?

File: pubsub/alicloud/mns/mns.go [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+	priority, _ := strconv.ParseInt(req.Metadata["priority"], 10, 64)
+	msgSendReq.Priority = priority
```
Comment (pkedy) [inline pubsub/alicloud/mns/mns.go:102]: Is `0` a valid value for `Priority`?

File: pubsub/alicloud/mns/mns.go [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+	addressType, _ := strconv.ParseInt(req.Metadata["AddressType"], 10, 32)
+	mailAttr.AddressType = int32(addressType)
```
Comment (pkedy) [inline pubsub/alicloud/mns/mns.go:105]: Is `0` a valid value for `AddressType`?

File: pubsub/alicloud/mns/mns.go [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+	replyToAddress, _ := strconv.ParseInt(req.Metadata["ReplyToAddress"], 10, 32)
+	mailAttr.ReplyToAddress = int32(replyToAddress)
```
Comment (pkedy) [inline pubsub/alicloud/mns/mns.go:108]: Is `0` a valid value for `ReplyToAddress`?

File: pubsub/alicloud/mns/mns.go [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+		if c, ok := m.subscriptionContexts[metaData.SubscriptionName]; ok {
+			c.cancel()
+		}
+		m.subscriptionContexts[metaData.SubscriptionName] = contextWithCancelFunc{ctx, cancelFunc}
+	}
+
+	return nil
+}
+
+func wrapMnsMessage(resp ali_mns.MessageReceiveResponse, topic string) (msg pubsub.NewMessage, err error) {
```
Comment (pkedy) [inline pubsub/alicloud/mns/mns.go:227]: Nit: Go naming `wrapMnsMessage` -> `wrapMNSMessage` (assuming MNS is an acronym)

File: pubsub/alicloud/mns/mns.go [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+	msg.Topic = topic
+	msg.Data = []byte(resp.MessageBody)
+
+	jsonBody, err := json.Marshal(msg)
+	if err != nil {
+		return msg, err
+	}
+
+	json.Unmarshal(jsonBody, &msg.Metadata)
```
Comment (pkedy) [inline pubsub/alicloud/mns/mns.go:236]: No error handling here. Should there be?

File: pubsub/alicloud/mns/mns.go [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+// Features does nothing here
+func (m *mns) Features() []pubsub.Feature {
+	return nil
+}
+
+// Close unsubscribes all topics/queues and closes this service gracefully
+func (m *mns) Close() error {
+	return nil
```
Comment (pkedy) [inline pubsub/alicloud/mns/mns.go:302]: I'm thinking you want a component level `ctx` and `cancel` so you can cleanly stop the goroutines started by the component.

**Review by pkedy — COMMENTED** [review]
Did a first pass review... mostly saw naming nitpicks.

**Comment by cshi0** [comment]
@pkedy Hi! I have fixed naming problems and the Close method. I wonder whether I need to check the validity of arguments if the sdk does it for me? The error will be returned by the Init method.

**Comment by yaron2** [comment]
hey @cshi0 can you resolve the conflicts?

@pkedy please look at the latest question from @cshi0

**Comment by cshi0** [comment]
> hey @cshi0 can you resolve the conflicts?
> 
> @pkedy please look at the latest question from @cshi0

Hi @yaron2 conflicts are resolved.

**Comment by pkedy** [comment]
> I wonder whether I need to check the validity of arguments if the sdk does it for me? The error will be returned by the Init method.

@cshi0 Do you mean the validity of the metadata properties? `settings.Decode(md.Properties)` returns an error that you need to check, but an error only happens when there is a type conversion error. You should also implement a `Validate` method on settings to perform validations at the field level.

Example:
https://github.com/dapr/components-contrib/blob/master/bindings/alicloud/nacos/settings.go

**Comment by cshi0** [comment]
> > I wonder whether I need to check the validity of arguments if the sdk does it for me? The error will be returned by the Init method.
> 
> @cshi0 Do you mean the validity of the metadata properties? `settings.Decode(md.Properties)` returns an error that you need to check, but an error only happens when there is a type conversion error. You should also implement a `Validate` method on settings to perform validations at the field level.
> 
> Example:
> https://github.com/dapr/components-contrib/blob/master/bindings/alicloud/nacos/settings.go

@pkedy The PublishMessage and SendMessage methods will complain if some metadata are out of the acceptable range.

File: pubsub/alicloud/mns/mns.go [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+package mns
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	ali_mns "github.com/aliyun/aliyun-mns-go-sdk"
+	"gopkg.in/square/go-jose.v2/json"
```
Comment (pkedy) [inline pubsub/alicloud/mns/mns.go:15]: Was this accidentally imported versus `encoding/json`?

**Comment by pkedy** [comment]
> @pkedy The PublishMessage and SendMessage methods called by Publish for sending the message to a queue or a topic will complain if some metadata are out of the acceptable range.

@cshi0 Right. So couldn't you implement those constraints using the `Validate` pattern from the example I linked to?

**Comment by cshi0** [comment]
> > @pkedy The PublishMessage and SendMessage methods called by Publish for sending the message to a queue or a topic will complain if some metadata are out of the acceptable range.
> 
> @cshi0 Right. So couldn't you implement those constraints using the `Validate` pattern from the example I linked to?

@pkedy I have added the validate pattern. Please run the test workflow again. Thanks!

**Comment by yaron2** [comment]
> > > @pkedy The PublishMessage and SendMessage methods called by Publish for sending the message to a queue or a topic will complain if some metadata are out of the acceptable range.
> > 
> > 
> > @cshi0 Right. So couldn't you implement those constraints using the `Validate` pattern from the example I linked to?
> 
> @pkedy I have added the validate pattern. Please run the test workflow again. Thanks!

Running.

**Comment by pkedy** [comment]
@cshi0 There are still a few outstanding questions. Could you please take a look?

**Comment by daixiang0** [comment]
ping @cshi0

**Review by daixiang0 — APPROVED** [review]

**Comment by yaron2** [comment]
@cshi0 please look at the failing build checks.

**Comment by cshi0** [comment]
> @cshi0 please look at the failing build checks.

@yaron2 The statement `CGO_ENABLED=0 go test ./...` does not work for windows, so I changed it.

File: pubsub/alicloud/mns/mns.go (hunk 14 above)
Comment (daixiang0) [reply pubsub/alicloud/mns/mns.go:15]: I notice that we use more than one json lib, one lib is enough I think.

/cc @artursouza @yaron2

**Comment by beiwei30** [comment]
@cshi0 would you mind to close the conversations below?


1. https://github.com/dapr/components-contrib/pull/962#discussion_r660129351
2. https://github.com/dapr/components-contrib/pull/962#discussion_r660129504
3. https://github.com/dapr/components-contrib/pull/962#discussion_r660129888
4. https://github.com/dapr/components-contrib/pull/962#discussion_r660130160
5. https://github.com/dapr/components-contrib/pull/962#discussion_r660137331

and resolve this question?


https://github.com/dapr/components-contrib/pull/962#discussion_r746146332

Thanks.

**Comment by beiwei30** [comment]
@cshi0 would you mind to make the lint clean? thanks.

**Comment by beiwei30** [comment]
ping @cshi0

**Pull request closed without merging**